hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+    fn set_raddr<T>(&mut self, addr: *const T);
+    /// Set the specific region configuration
+    fn set_rcfg(&mut self, cfg: Rcfg);
+    /// Set the link to next region descriptor
+    fn set_rnext(&mut self, next: Rnext);
+    /// Reset Rcfg to default values
+    fn reset_rcfg_to_default(&mut self);
+}
+
+/// Helper for creating the Region Descriptor structure
+///
+/// It is also possible to construct the IcmRegion manually,
+/// but then care has to be taken to point `rnext` to the appropriate
+/// place in memory, here the hardware assumption of 0x10
+/// offset to the next descriptor is ensured.
+///
+/// Follows C-structure conventions and is 64-byte aligned
+///
+/// >**Important**
+/// >
+/// >ICM engine will **read** wherever this
+/// >is instantiated in memory, based on the [`Icm::set_dscr_addr()`]
+/// >so the user must ensure that this variable lives long enough or is
+/// >static
+#[repr(C)]
+#[repr(align(64))]
+pub struct IcmRegion {
+    /// IcmRegionDesc0
+    pub region0: IcmRegionDesc<Region0>,
+    /// IcmRegionDesc1
+    pub region1: IcmRegionDesc<Region1>,
+    /// IcmRegionDesc2
+    pub region2: IcmRegionDesc<Region2>,
+    /// IcmRegionDesc3
+    pub region3: IcmRegionDesc<Region3>,
+}
+
+impl IcmRegion {
+    pub const fn default() -> Self {
+        let region0 = IcmRegionDesc::new_region0();
+        let region1 = IcmRegionDesc::new_region1();
+        let region2 = IcmRegionDesc::new_region2();
+        let region3 = IcmRegionDesc::new_region3();
+        Self {
+            region0,
+            region1,
+            region2,
+            region3,
+        }
+    }
+}
+
+/// Structure ICM Region Descriptor area.
+///
+/// Follows C-structure conventions and is 16-byte aligned,
+/// being a part of the 64-bytes making up [`IcmRegion`]
+#[repr(C)]
+#[repr(align(16))]
+pub struct IcmRegionDesc<N: IcmRegionNum> {
+    /// Numerical Region Identifier
+    num: PhantomData<N>,
+    /// The first byte address of the Region.
+    pub raddr: Raddr,
+    /// Configuration Structure Member.
+    pub rcfg: Rcfg,
+    /// Control Structure Member.
+    pub rctrl: Rctrl,
+    /// Next Address Structure Member.
+    pub rnext: Rnext,
+}
+
+impl IcmRegionDesc<Region0> {
+    /// Helper for setting the DSCR addr to a the first IcmRegionDesc
+    ///
+    /// See [`Icm::set_dscr_addr()`] for the regular workflow
+    ///
+    /// HW expects a raw pointer to the memory address of the beginning of the
+    /// [`IcmRegionDesc`] but expressed as a multiple of 64
+    pub fn set_dscr_addr(&self, icm: &Icm) {
+        icm.token
+            .dscr()
+            .write(|w| unsafe { w.dasa().bits((self as *const _) as u32 / 64) })
+    }
+
+    const fn new_region0() -> Self {
+        IcmRegionDesc {
+            num: PhantomData,
+            raddr: Raddr::default(),
+            rcfg: Rcfg::default(),
+            rctrl: Rctrl::default(),
+            rnext: Rnext::default(),
+        }
+    }
```

Comment:
Why not pull this out into a generic version?

```rust
impl<R: RegionNum> MainRegionDesc<R> {
    const fn new_region() -> Self {
        // ...
    }
}
```

If you really want separate names, you can then use a `seq!` macro to create the four different versions.

Reply:
Also, why `new_region` instead of just `new`? Doesn't the type name already imply it's a region?